src/components/TodosContext.tsx

```diff
@@ -0,0 +1,205 @@
+import React, {
+  Dispatch,
+  SetStateAction, useCallback, useMemo, useRef, useState,
+} from 'react';
+import { Status, Todo } from '../types/Todo';
+import * as postService from '../api/todos';
+
+type TodosContextType = {
+  todos: Todo[];
+  setTodos: Dispatch<SetStateAction<Todo[]>>;
+  selectedStatus: Status;
+  setSelectedStatus: Dispatch<SetStateAction<Status>>;
+  title: string,
+  setTitle: Dispatch<SetStateAction<string>>,
+  visibleTodos: Todo[],
+  loadingTodos: boolean,
+  setLoadingTodos: Dispatch<SetStateAction<boolean>>,
+  errorMessage: string,
+  setErrorMessage: Dispatch<SetStateAction<string>>,
+  errorHidden: boolean,
+  setErrorHidden: Dispatch<SetStateAction<boolean>>,
+  tempTodo: Todo | null,
+  setTempTodo: Dispatch<SetStateAction<Todo | null>>,
+  loading: boolean,
+  setLoading: Dispatch<SetStateAction<boolean>>,
+  selectedTodo: Todo | null,
+  setSelectedTodo: Dispatch<SetStateAction<Todo | null>>,
+  deleteTodo: (todoId: number) => void,
+  updateTodo: (todo: Todo) => void,
+  completedTodoIds: number[],
+  deleteLoading: boolean,
+  setDeleteLoading: Dispatch<SetStateAction<boolean>>,
+  toggleAllLoading: boolean,
+  setToggleAllLoading: Dispatch<SetStateAction<boolean>>,
+  inputRef: React.RefObject<HTMLInputElement> | null,
+  focusInput: () => void,
+};
+
+export const TodosContext = React.createContext<TodosContextType>({
+  todos: [],
+  setTodos: () => {},
+  selectedStatus: Status.all,
+  setSelectedStatus: () => {},
+  title: '',
+  setTitle: () => {},
+  visibleTodos: [],
+  loadingTodos: false,
+  setLoadingTodos: () => {},
+  errorMessage: '',
+  setErrorMessage: () => {},
+  errorHidden: true,
+  setErrorHidden: () => {},
+  tempTodo: null,
+  setTempTodo: () => {},
+  loading: false,
+  setLoading: () => {},
+  selectedTodo: null,
+  setSelectedTodo: () => {},
+  deleteTodo: () => {},
+  updateTodo: () => {},
+  completedTodoIds: [],
+  deleteLoading: false,
+  setDeleteLoading: () => {},
+  toggleAllLoading: false,
+  setToggleAllLoading: () => {},
+  inputRef: null,
+  focusInput: () => {},
+});
+
+type Props = {
+  children: React.ReactNode,
+};
+
+export const TodosProvider: React.FC<Props> = ({ children }) => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [selectedStatus, setSelectedStatus] = useState(Status.all);
+  const [title, setTitle] = useState('');
+
+  const [loadingTodos, setLoadingTodos] = useState(false);
+  const [errorMessage, setErrorMessage] = useState('');
+  const [errorHidden, setErrorHidden] = useState(true);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [loading, setLoading] = useState(false);
+  const [selectedTodo, setSelectedTodo] = useState<Todo | null>(null);
+  const [deleteLoading, setDeleteLoading] = useState(false);
+  const [toggleAllLoading, setToggleAllLoading] = useState(false);
+
+  const visibleTodos = [...todos].filter(todo => {
```

Comment:
```suggestion
  const visibleTodos = todos.filter(todo => {
```
the `filter` method doesn't mutate the array